torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,293 @@
+      options = options.dtype(scalar_type_);
+    }
+
+    if (is_scalar()) {
+      at::AutoNonVariableTypeMode non_var_type_mode(true);
+      return at::scalar_tensor(scalar_, options.is_variable(false));
+    } else if (is_init_list()) {
+      // NOTE: Here we explicitly choose to initialize the tensor on CPU first,
+      // fill each element of the tensor, and then move the tensor to the desired
+      // device. For CUDA device, this approach only involves 1 CUDA kernel launch,
+      // and is much faster than initializing the tensor on CUDA first and then
+      // filling each element of it (which involves `N` CUDA kernel launches where
+      // `N` is the number of the elements in the tensor).
+      at::Tensor tensor = ([&]() {
+        at::AutoNonVariableTypeMode non_var_type_mode(true);
+        return at::empty(sizes_, options.device(at::kCPU).is_variable(false));
+      })();
+      fill_tensor(tensor);
+      return tensor.to(options.device());
+    } else if (is_tensor()) {
+      return tensor_.to(options);
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "Invalid TensorDataContainer type");
+    }
+  }
+
+  void pretty_print_recursive(std::ostream& stream) const {
+    if (is_scalar()) {
+      AT_DISPATCH_ALL_TYPES_AND3(
+          at::kBool,
+          at::kHalf,
+          at::kBFloat16,
+          scalar_type_,
+          "TensorDataContainer_pretty_print_scalar", [&] {
+        stream << scalar_.to<scalar_t>();
+      });
+    } else if (is_init_list()) {
+      stream << "{";
+      for (const TensorDataContainer* it = init_list_.begin(); it != init_list_.end(); it++) {
+        stream << *it;
+        if (std::next(it) != init_list_.end()) stream << ", ";
+      }
+      stream << "}";
+    } else if (is_tensor()) {
+      stream << "{";
+      for (int64_t i = 0; i < tensor_.sizes()[0]; i++) {
+        AT_DISPATCH_ALL_TYPES_AND3(
+            at::kBool,
+            at::kHalf,
+            at::kBFloat16,
+            scalar_type_,
+            "TensorDataContainer_pretty_print_tensor_item", [&] {
+          stream << tensor_[i].item<scalar_t>();
+        });
+        if (i != tensor_.sizes()[0] - 1) stream << ", ";
+      }
+      stream << "}";
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "Invalid TensorDataContainer type");
+    }
+  }
+ private:
+  void fill_tensor(at::Tensor& tensor) const {
+    if (is_scalar()) {
+      TORCH_INTERNAL_ASSERT(
+        tensor.dim() == 0,
+        "Expected a 0-dim Tensor, but got Tensor with dimensions: ", tensor.dim());
+      at::NoGradGuard guard;
+      tensor.fill_(scalar_);
+    } else if (is_init_list()) {
+      TORCH_INTERNAL_ASSERT(
+        tensor.sizes()[0] == (int64_t)init_list_.size(),
+        "Expected a Tensor with size ",
+        init_list_.size(),
+        " in its first dimension, but got Tensor with size ",
+        tensor.sizes()[0],
+        " in its first dimension");
+      size_t index = 0;
+      for (const auto& elem : init_list_) {
```

Comment:
Does std::initializer_list support indexing? if so it looks like it might be nicer as a for loop: `for (size_t index = 0; index < init_list_.size(); ++index)`

Reply:
I thought about the same too, but the author of C++ said it doesn't support subscripting :( https://stackoverflow.com/a/17823064